pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -660,8 +664,51 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 // for the passed-in <pod, container> and returns its DeviceRunContainerOptions
 // for the found one. An empty struct is returned in case no cached state is found.
 func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
+	podUID := string(pod.UID)
+	contName := container.Name
+	for k, v := range container.Resources.Limits {
+		resource := string(k)
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, resource)
+		_, registeredResource := m.healthyDevices[resource]
+		_, allocatedResource := m.allocatedDevices[resource]
+		// Continues if this is neither an active device plugin resource nor
+		// a resource we have previously allocated.
+		if !registeredResource && !allocatedResource {
+			continue
+		}
+		m.mutex.Lock()
+		e, ok := m.endpoints[resource]
+		if !ok {
+			m.mutex.Lock()
+			m.allocatedDevices = m.podDevices.devices()
+			m.mutex.Unlock()
+			glog.Errorf("Unknown Device Plugin %s", resource)
+			return nil
+		}
+		m.mutex.Unlock()
+		if e.getInitReqFlag() {
+			glog.V(3).Infof("getInitReqFlag is true")
+			m.mutex.Lock()
+			devices := m.podDevices.containerDevices(podUID, contName, resource)
+			m.mutex.Unlock()
+			if devices == nil {
+				glog.Errorf("No devices found allocated in local cache for pod %s, container %s, resource %s", podUID, contName, resource)
+				m.mutex.Unlock()
```

Comment:
Why the unlock here ?

This whole high level mutex complexity here illustrates one of the reasons for having a have a per component mutex approach not a "lock the world" one.

A big advantage of having mutexes protecting only the data would allow for a defer Unlock pattern preventing deadlocks and mutex confusings.

Reply:
This one is left by mistake and not required. Will remove it, thanks!!!